src/server/v2.0/handler/project.go

```diff
@@ -63,3 +241,380 @@ func (a *projectAPI) GetLogs(ctx context.Context, params operation.GetLogsParams
+	for _, p := range projects {
+		wg.Add(1)
+		go func(p *project.Project) {
+			defer wg.Done()
+			// due to the issue https://github.com/lib/pq/issues/81 of lib/pg or postgres,
+			// simultaneous queries in transaction may failed, so clone a ctx with new ormer here
+			if err := a.populateProperties(orm.Clone(ctx), p); err != nil {
+				log.G(ctx).Errorf("failed to populate propertites for project %s, error: %v", p.Name, err)
+			}
+		}(p)
+	}
+	wg.Wait()
+
+	var payload []*models.Project
+	for _, p := range projects {
+		payload = append(payload, model.NewProject(p).ToSwagger())
+	}
+
+	return operation.NewListProjectsOK().
+		WithXTotalCount(total).
+		WithLink(a.Links(ctx, params.HTTPRequest.URL, total, query.PageNumber, query.PageSize).String()).
+		WithPayload(payload)
+}
+
+func (a *projectAPI) UpdateProject(ctx context.Context, params operation.UpdateProjectParams) middleware.Responder {
+	if err := a.RequireProjectAccess(ctx, params.ProjectID, rbac.ActionUpdate); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	p, err := a.projectCtl.Get(ctx, params.ProjectID, project.Metadata(false))
+	if err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	lib.JSONCopy(&p.Metadata, params.Project.Metadata)
+
+	if params.Project.CVEAllowlist != nil {
+		// force the project id of the cve allowlist to be the project which will be update
+		params.Project.CVEAllowlist.ProjectID = params.ProjectID
+		lib.JSONCopy(&p.CVEAllowlist, params.Project.CVEAllowlist)
+	}
+
+	if err := a.projectCtl.Update(ctx, p); err != nil {
+		return a.SendError(ctx, err)
+	}
+
+	return operation.NewUpdateProjectOK()
+}
+
+func (a *projectAPI) deletable(ctx context.Context, projectID int64) (*models.ProjectDeletable, error) {
+	proj, err := a.getProject(ctx, projectID)
+	if err != nil {
+		return nil, err
+	}
+
+	result := &models.ProjectDeletable{Deletable: true}
+	if proj.RepoCount > 0 {
+		result.Deletable = false
+		result.Message = "the project contains repositories, can not be deleted"
+	} else if proj.ChartCount > 0 {
+		result.Deletable = false
+		result.Message = "the project contains helm charts, can not be deleted"
+	}
+
+	return result, nil
+}
+
+func (a *projectAPI) getProject(ctx context.Context, projectID int64, options ...project.Option) (*project.Project, error) {
+	p, err := a.projectCtl.Get(ctx, projectID, options...)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := a.populateProperties(ctx, p); err != nil {
+		return nil, err
+	}
+
+	return p, nil
+}
+
+func (a *projectAPI) validateProjectReq(ctx context.Context, req *models.ProjectReq) error {
+	if req.RegistryID != nil {
+		if *req.RegistryID <= 0 {
+			return errors.BadRequestError(fmt.Errorf("%d is invalid value of registry_id, it should be geater than 0", *req.RegistryID))
+		}
+
+		registry, err := replication.RegistryMgr.Get(*req.RegistryID)
+		if err != nil {
+			return fmt.Errorf("failed to get the registry %d: %v", *req.RegistryID, err)
+		}
+		if registry == nil {
```

Comment:
we should update RegistryMgr.Get to give a 404, but not in this PR.

Reply:
Yes, this is the scope of refactoring the manager of the registry.